src/gui/PianoView.cpp

```diff
@@ -502,7 +544,52 @@ void PianoView::mousePressEvent(QMouseEvent *me)
 	}
 }
 
+bool PianoView::setMarkerKeyValue(IntModel *noteModel, int key_num, bool ignoreConstraints)
+{
+	if (!ignoreConstraints)
+	{
+		if (m_piano->instrumentTrack()->firstKeyModel() == noteModel && !firstMarkerAllowed(key_num)) return false;
+		if (m_piano->instrumentTrack()->lastKeyModel() == noteModel && !lastMarkerAllowed(key_num)) return false;
+	}
+	noteModel->setValue(static_cast<float>(key_num));
+	if (noteModel == m_piano->instrumentTrack()->baseNoteModel()) { emit baseNoteChanged(); }	// TODO: not actually used by anything?
+	return true;
+}
 
+bool PianoView::firstMarkerAllowed(int key_num)
+{
+	return key_num <=  m_piano->instrumentTrack()->lastKeyModel()->value();
+}
+
+bool PianoView::lastMarkerAllowed(int key_num)
+{
+	return key_num >=  m_piano->instrumentTrack()->firstKeyModel()->value();
+}
+
+
+void PianoView::setBaseNote()
+{
+	setMarkerKeyValue(m_piano->instrumentTrack()->baseNoteModel(), m_lastContextMenuKey);
+	update();
+}
+
+void PianoView::setFirstKey()
+{
+	if (setMarkerKeyValue(m_piano->instrumentTrack()->firstKeyModel(), m_lastContextMenuKey)) update();
```

Comment:
The updates here are not necessary as they will be triggered by

    connect(m_piano->instrumentTrack()->firstKeyModel(), SIGNAL(dataChanged()), this, SLOT(update()));